admin/python/lsst/qserv/testing/config.py

```diff
@@ -0,0 +1,299 @@
+
+    Parameters
+    ----------
+    path_or_file : `str` or file object
+        Path name for a file or a file object.
+    """
+    if hasattr(path_or_file, "read"):
+        yield path_or_file
+    else:
+        with open(path_or_file) as file:
+            yield file
+
+
+class _ValueRandomUniform:
+    def __init__(self, min, max):
+        self._min = float(min)
+        self._max = float(max)
+
+    def __call__(self):
+        return random.uniform(self._min, self._max)
+
+
+class _ValueRandomUniformInt:
+    def __init__(self, min, max):
+        self._min = float(min)
+        self._max = float(max)
+
+    def __call__(self):
+        return int(random.uniform(self._min, self._max))
+
+
+class _ValueIntFromFile:
+    def __init__(self, path, mode="random"):
+        # read all numbers from file as integers
+        if path == "/dev/null":
+            # for testing only
+            self._array = [0]
+        else:
+            self._array = np.fromfile(path, dtype=int, sep=" ")
+        self._mode = mode
+        assert mode in ("random", "sequential")
+        self._seq = 0
+
+    def __call__(self):
+        if self._mode == "random":
+            return random.choice(self._array)
+        else:
+            if self._seq >= len(self._array):
+                self._seq = 0
+            value = self._array[self._seq]
+            self._seq += 1
+            return value
+
+
+class QueryFactory:
+    """Class which generates queries.
+
+    Parameters
+    ----------
+    txt : `str`
+        If ``variables`` is ``None`` then full query text, otherwise it is a
+        query template.
+    variables : `dict` [`str`, `dict`]
+        Dictionary whose keys are variable names and values are dictionaries
+        with a description of how to generate variable value.
+    """
+    def __init__(self, txt, variables=None):
+        self._txt = txt
+        self._vars = {}
+        if variables is not None:
+
+            for var, config in variables.items():
+                generator = None
+                if "distribution" in config:
+                    if config["distribution"] == "uniform":
+                        min = config.get("min", 0.)
+                        max = config.get("max", 1.)
+                        generator = _ValueRandomUniform(min, max)
+                    elif config["distribution"] == "uniform_int":
+                        min = config.get("min", 0)
+                        max = config.get("max", 2**64)
+                        generator = _ValueRandomUniformInt(min, max)
+                elif "path" in config:
+                    path = config["path"]
+                    mode = config.get("mode", "random")
+                    generator = _ValueIntFromFile(path, mode)
+                if generator is None:
+                    raise ValueError(f"Cannot parse variable configuration {var} = {config}")
+                self._vars[var] = generator
+
+    def query(self):
```

Comment:
pls add docstring with Returns doc

Reply:
my specific q is, will the return always be a dict? if it returns self._txt could it be a string that is not in a container?

Reply:
It always returns string - full text of the query. We interpolate query variables directly into the query text, not passing them to DBAPI level. Potentially we could pass variables as separate dictionary, for MySQL (or for how we uset) that does not really matter.